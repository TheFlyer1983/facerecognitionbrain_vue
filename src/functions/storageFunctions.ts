export const saveAuthTokenInSession = (token: string): void => {
  window.sessionStorage.setItem('token', token);
};

export const getAuthTokenInSession = () => {
  return window.sessionStorage.getItem('token');
}