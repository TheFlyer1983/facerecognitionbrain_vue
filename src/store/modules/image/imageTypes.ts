export interface ImageState {
  input: string;
  imageUrl: string;
  boxes: Array<Box>;
}

export interface Box {
  id: string;
  region_info: Region;
  data: Array<Concept>;
  value: number;
}

export interface Region {
  bounding_box: BoundingBox;
}

export interface BoundingBox {
  top_row: number;
  left_col: number;
  bottom_row: number;
  right_col: number;
}

export interface Concept {
  id: string;
  name: string;
  value: number;
  app_id: string;
}
