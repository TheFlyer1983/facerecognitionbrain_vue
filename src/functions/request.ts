import axios from 'axios';
import { getAuthTokenInSession } from './storageFunctions';

const client = axios.create();

client.interceptors.request.use(async (config) => {
  const token = getAuthTokenInSession();

  if (!token) return config;

  if (config.headers) {
    config.headers.Authorization = token
  }

  return config;
});

client.interceptors.response.use(
  (response) => response,
  async (error) => {
    return Promise.reject(error.message);
  }
);

export default client;
