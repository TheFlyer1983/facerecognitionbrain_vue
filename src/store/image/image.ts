import { defineStore } from 'pinia';
import { useUserStore } from '../user';
import request from '@/functions/request';
import { ImageState } from './imageTypes';
import { endpoints } from '@/constants';

export const useImageStore = defineStore('image', {
  state: (): ImageState => ({
    imageUrl: '',
    boxes: []
  }),

  actions: {
    async submitURL() {
      try {
        const response = await request.post(endpoints.imageURL, {
          input: this.imageUrl
        });

        this.boxes = response.data;
        this.increaseEntries();
      } catch (error) {
        console.error(error);
      }
    },

    async increaseEntries() {
      const userStore = useUserStore();

      if (!userStore.user) return;

      try {
        const response = await request.put(endpoints.image, {
          id: userStore.user.id
        });

        userStore.user.entries = response.data.entries;
        userStore.getRank();
      } catch (error) {
        console.error(error);
      }
    }
  }
});
