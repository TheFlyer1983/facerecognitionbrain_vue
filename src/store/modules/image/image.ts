import { endpoints } from '@/constants';
import { UserState } from '../user/userTypes';
import { ImageActionContext, ImageState } from './imageTypes';
import request from '@/functions/request';

export const state: ImageState = {
  imageUrl: '',
  boxes: []
};

export const getters = {
  getImageURL(state: ImageState): ImageState['imageUrl'] {
    return state.imageUrl;
  },

  getBoxes(state: ImageState): ImageState['boxes'] {
    return state.boxes;
  }
};

export const mutations = {
  setImageURL(state: ImageState, payload: ImageState['imageUrl']): void {
    state.imageUrl = payload;
  },

  setBoxes(state: ImageState, payload: ImageState['boxes']): void {
    state.boxes = payload;
  }
};

export const actions = {
  async submitURL({
    state,
    commit,
    dispatch
  }: ImageActionContext): Promise<void> {
    try {
      const response = await request.post(endpoints.imageURL, {
          input: state.imageUrl
        });

      commit('setBoxes', response.data);
      dispatch('increaseEntries');
    } catch (error) {
      console.error(error);
    }
  },

  async increaseEntries({
    commit,
    dispatch,
    rootGetters
  }: ImageActionContext): Promise<void> {
    const user = rootGetters['user/getUser'] as UserState['user'];

    try {
      const response = await request.put(endpoints.image, {
          id: user?.id
        });

      commit('user/updateEntries', response.data.entries, { root: true });
      dispatch('user/getRank', null, { root: true });
    } catch (error) {
      console.error(error);
    }
  }
};

const imageModule = {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
};

export default imageModule;
