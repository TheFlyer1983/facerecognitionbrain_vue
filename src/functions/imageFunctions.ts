import { Box } from '@/store/image/imageTypes';
import { FaceLocation } from '@/types';

export const calculateFaceLocations = (
  data: Array<Box>
): Array<FaceLocation> => {
  const image = document.getElementById('inputImage') as HTMLImageElement;
  const width = image.width;
  const height = image.height;

  return data.map((face) => {
    const clarifaiFace = face.region_info.bounding_box;
    return {
      leftCol: clarifaiFace.left_col * width,
      topRow: clarifaiFace.top_row * height,
      rightCol: width - clarifaiFace.right_col * width,
      bottomRow: height - clarifaiFace.bottom_row * height
    };
  });
};

export const createFaceRecognitionPayload = (imageUrl: string) => ({
  user_app_id: {
    user_id: 'clarifai',
    app_id: 'main'
  },
  inputs: [
    {
      data: {
        image: {
          url: imageUrl
        }
      }
    }
  ]
});
