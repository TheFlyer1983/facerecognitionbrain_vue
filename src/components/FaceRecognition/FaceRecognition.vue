<script setup lang="ts">
import { computed } from 'vue';
import { storeToRefs } from 'pinia';
import { useImageStore } from '@/store/image';
import { Box } from '@/store/image/imageTypes';
import { calculateFaceLocations } from '@/functions/imageFunctions';

const imageStore = useImageStore();

const { imageUrl } = storeToRefs(imageStore);

const boxes = computed(() => {
  const data: Array<Box> = imageStore.boxes;
  if (data.length) {
    return calculateFaceLocations(data);
  }
});
</script>

<template>
  <div class="face-recognition">
    <div class="face-recognition-wrapper">
      <img
        id="inputImage"
        :src="imageUrl"
        alt=""
        width="500"
        class="face-recognition-wrapper--image"
      />
      <div
        v-for="box in boxes"
        :key="box.topRow"
        class="face-recognition-wrapper--bounding_box"
        :style="`inset: ${box.topRow}px ${box.rightCol}px ${box.bottomRow}px ${box.leftCol}px`"
      ></div>
    </div>
  </div>
</template>

<style scoped lang="scss">
.face-recognition {
  display: flex;
  justify-content: center;

  &-wrapper {
    position: absolute;
    margin-top: var(--spacing-small);

    &--bounding_box {
      position: absolute;
      box-shadow: 0 0 0 3px #149df2 inset;
      display: flex;
      flex-wrap: wrap;
      justify-content: center;
      cursor: pointer;
    }
  }
}
</style>
