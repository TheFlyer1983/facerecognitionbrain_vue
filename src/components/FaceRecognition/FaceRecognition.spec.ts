import { describe, it, expect, vi, beforeAll, beforeEach } from 'vitest';
import { shallowMount, VueWrapper } from '@vue/test-utils';
import { reactive } from 'vue';
import { useStore } from 'vuex';

import FaceRecognition from './FaceRecognition.vue';
import { ImageState } from '@/store/modules/image/imageTypes';
import { boxesMock } from '@/fixtures/images';
import { calculateFaceLocations } from '@/functions/imageFunctions';

vi.mock('vuex');
vi.mock('@/functions/imageFunctions');

const mockedUseStore = vi.mocked<() => Partial<typeof useStore>>(useStore);
const mockedCalculateFaceLocations = vi.mocked(calculateFaceLocations, true);

const mockedImageURL =
  'https://images.unsplash.com/photo-1597223557154-721c1cecc4b0?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MjB8fGZhY2V8ZW58MHx8MHx8&w=1000&q=80';

const mockedFaceLocation = [
  {
    leftCol: 105.42601499999999,
    topRow: 74.074805,
    rightCol: 122.83062999999999,
    bottomRow: 44.97919999999999
  }
];

describe('Given the `FaceRecognition` component', () => {
  const render = () => shallowMount(FaceRecognition);
  let wrapper: VueWrapper<InstanceType<typeof FaceRecognition>>;

  const mockedGetters = {
    ['image/getImageURL']: mockedImageURL as ImageState['imageUrl'],
    ['image/getBoxes']: boxesMock as ImageState['boxes']
  };

  const mockedStore = reactive({
    getters: mockedGetters
  });

  beforeAll(() => {
    mockedUseStore.mockReturnValue(mockedStore);
    mockedCalculateFaceLocations.mockReturnValue(mockedFaceLocation);
  });

  describe('when it is rendered', () => {
    beforeEach(() => {
      wrapper = render();
    });

    it('should match the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
