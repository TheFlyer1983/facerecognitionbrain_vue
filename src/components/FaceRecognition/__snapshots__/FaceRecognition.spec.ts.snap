// Vitest Snapshot v1

exports[`Given the \`ImageLinkForm\` component > when it is rendered > should match the snapshot 1`] = `
<div
  class="face-recognition"
  data-v-475b3985=""
>
  <div
    class="face-recognition-wrapper"
    data-v-475b3985=""
  >
    <img
      alt=""
      class="face-recognition-wrapper--image"
      data-v-475b3985=""
      id="inputImage"
      src="https://images.unsplash.com/photo-1597223557154-721c1cecc4b0?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxzZWFyY2h8MjB8fGZhY2V8ZW58MHx8MHx8&w=1000&q=80"
      width="500"
    />
    
    <div
      class="face-recognition-wrapper--bounding_box"
      data-v-475b3985=""
      style="inset: 74.074805px 122.83062999999999px 44.97919999999999px 105.42601499999999px;"
    />
    
  </div>
</div>
`;
